lib/oxigraph/src/sparql/dataset.rs

```diff
@@ -1,18 +1,235 @@
+/// or blank node) the string has. Based on
+/// https://github.com/KonradHoeffner/hdt/blob/871db777db3220dc4874af022287975b31d72d3a/src/hdt_graph.rs#L64
+fn hdt_bgp_str_to_term(s: &str) -> Result<Term, Error> {
+    match s.chars().next() {
+        None => Err(Error::new(ErrorKind::InvalidData, "empty input")),
+
+        // Double-quote delimters are used around the string.
+        Some('"') => match s.rfind('"') {
+            None => Err(Error::new(
+                ErrorKind::InvalidData,
+                format!("missing right quotation mark in literal string {s}"),
+            )),
+
+            Some(_) => match Term::from_str(s) {
+                Ok(s) => Ok(s),
+                Err(e) => Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("literal parse error {e} for {s}"),
+                )),
+            },
+        },
+
+        // Underscore prefix indicating an Blank Node.
+        Some('_') => {
+            let term = match oxrdf::BlankNode::new(&s[2..]) {
+                Ok(n) => n,
+                Err(e) => {
+                    return Err(Error::new(
+                        ErrorKind::InvalidData,
+                        format!("blanknode parse error {e} for {s}"),
+                    ))
+                }
+            };
+            match Term::from_str(&term.to_string()) {
+                Ok(s) => Ok(s),
+                Err(e) => Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("blanknode parse error {e} for {s}"),
+                )),
+            }
+        }
+
+        // Double-quote delimiters not present. Underscore prefix
+        // not present. Assuming a URI.
+        _ => {
+            // Note that Term::from_str() will not work for URIs
+            // (OxRDF NamedNode) when the string is not within "<"
+            // and ">" delimiters.
+            let named_node = match NamedNodeRef::new(*Arc::from(s)) {
+                Ok(n) => n,
+                Err(e) => {
+                    return Err(Error::new(
+                        ErrorKind::InvalidData,
+                        format!("iri parse error {e} for {s}"),
+                    ))
+                }
+            };
+            match Term::from_str(&named_node.to_string()) {
+                Ok(s) => Ok(s),
+                Err(e) => Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("iri parse error {e} for {s}"),
+                )),
+            }
+        }
+    }
+}
+
+/// Convert triple string formats from OxRDF to HDT.
+fn term_to_hdt_bgp_str(term: &Term) -> Result<String, StorageError> {
+    let hdt_str = match term {
+        // Remove double quote delimiters from URIs.
+        Term::NamedNode(named_node) => named_node.clone().into_string(),
+
+        // Get the string directly from literals and add
+        // quotes to work-around handling of "\n" being
+        // double-escaped.
```

Comment:
The `\n` unescaping is scary because oxrdf escape other characters (", \r...) and unescape them in `Literal::from_str`

Reply:
Yes, indeed.
It seems hdt-cpp is missing some escapes that are required for proper NTriples. For example data imported from Turtle are not escaped at all: https://github.com/rdfhdt/hdt-cpp/blob/d9ae092bb37d9fe85558dfb3edfe0bb6ddddf41a/libhdt/src/rdf/RDFParserSerd.cpp#L46

This might cause weird bugs for string containing eg. `"`.